Extract cache impls to their own packages

cies opened the pull request:
Closes #454

xabolcs commented on framework/src/play/cache/Cache.java:
```diff
@@ -63,17 +66,25 @@ public static void clear() {
+             | InvocationTargetException
+             | InstantiationException
+             | IllegalAccessException e) {
+      cache = new DummyCacheImpl();
```
```suggestion
      cache = new DummyCacheImpl();
      logger.info("Using dummy cache (no caching)");
```

cies replied:
I prefer not to. It's already waaaaaaaay too chatty at startup for my taste. I know there's no caching "in RePlay" as I did not set it up. If I need it I'll learn about it.

But we have a ton of caching still in many other places: hibernate, redis, jvm, CPU... It's not very clear what this "no caching" means. So I rather do without.

asolntsev replied:
@cies The problem is when I **want** to use cache, and I add "ehcache" or "memcached" module. BUT `ClassNotFoundException` happens. Then I definitely need to see some error in log to understand what was the problem.

cies replied:
@asolntsev Sorry to've bothered you with the review: I had some insights and found that the code I submitted for review was bad in so many aspects. I've just pushed a new attempt, that fixes all feedback, uses my new understandings and adds tests.

xabolcs commented on memcached/build.gradle:
```diff
@@ -0,0 +1,8 @@
+  testImplementation project(':framework').sourceSets.test.compileClasspath
+}
+
+apply from: rootProject.file('gradle/deploy.gradle')
```
No new line at end of file.

`.editorconfig` is not active yet?

cies replied:
what i'd do: i just run formatter once in a while on `main`.

but I'll fix it for you. (prolly will never develop into a habit though)

cies replied:
Fixed!

asolntsev commented on ehcache/test/play/cache/EhCacheImplTest.java:
```diff
@@ -154,7 +154,7 @@ public void offHeapStorageWorks() {
     assertThat(resourcePools.getPoolForResource(OFFHEAP).getSize()).isEqualTo(1);
 
-    cache.set("test", 1, 1);
+    locaCache.set("test", 1, 1);
```
typo: "locaCache"

cies replied:
oops. fixed it.

asolntsev commented on memcached/src/play/cache/MemcachedImpl.java:
```diff
@@ -40,6 +41,14 @@ public MemcachedClient getClient() {
+      uniqueInstance = new MemcachedImpl(playProperties);
+    }
+    return null;
+  }
```
I don't understand this place.
1. non-static method in `MemcachedImpl` returns another instance of `MemcachedImpl`? 
2. This method always returns `null`.

cies replied:
As mentioned in the other review-comment-thread, the code has been greatly improved now. Also due to this comment.

asolntsev commented on ehcache/src/play/cache/EhCacheImpl.java:
```diff
@@ -62,10 +63,14 @@ private EhCacheImpl() {
+  public EhCacheImpl instance(@Nonnull Properties playProperties) {
+    if (uniqueInstance == null) {
+      uniqueInstance = new EhCacheImpl();
+    }
```
I don't understand this place.
Non-static method in `EhCacheImpl` returns another instance of `EhCacheImpl`?

cies replied:
Fixed.

xabolcs commented:
Commit d2f672bfe2c05a68ce2f757e9d92cceb3cc8c0d1 ("Improve docs") should be dropped (from here, a reworked patch could be resend with another PR), IMHO:
* not related to this cache refactor
* the `'` -> `"` changes in the strings are really questionable anyway: in Groovy if you don't use a variable in the string then it's better to use `'`

@asolntsev please squash this once get merged!
`main` branch doesn't need to know about how this PR evolved. For those who are interested in it, just check this PR!

cies commented:
> Commit [d2f672b](https://github.com/replay-framework/replay/commit/d2f672bfe2c05a68ce2f757e9d92cceb3cc8c0d1) ("Improve docs") should be dropped (from here, a reworked patch could be resend with another PR), IMHO:
> 
>     * not related to this cache refactor

I have also moved the line that refs the ehcache package more down (in the Optional section where it belongs). I did this as I promised you to add to the docs how to use the different caching systems in RePlay.

Personally, I dont care much if small fixes are done along side changes that relate to an issue. I call that "good boyscouting" and I promote it.

I'm not the only one, lots of articles on this, like: https://medium.com/engineering-managers-journal/the-boy-scout-rule-for-better-code-e6ffda090902

Sure I understand this practice becomes a problem in really large teams with code that's constantly changed by different teams. RePlay is no such project.

>     * the `'` -> `"` changes in the strings are really questionable anyway: in Groovy if you don't use a variable in the string then it's better to use `'`

Now it is valid Groovy and valid KTS. That's why I changed it.

> Please squash this once get merged! `main` branch doesn't need to know about how this PR evolved.

Will do if I merge it. (and not mess it up again)

xabolcs commented:
> I have also moved the line that refs the ehcache package more down (in the Optional section where it belongs). I did this as I promised you to add to the docs how to use the different caching systems in RePlay.

@cies 
Ouch! 

Now I see, as you mentioned! :sweat_smile: 
Sorry! You are right! And thanks for adding an example.

I hope now you also see why am I asking about this "one thing at a time": I just missed that one relevant line from the 21 changed lines. :shrug: 
And yeah, the commit message didn't help.


If you drop those reformat changes from README, but keeping just adding the example that would be awesome.
But I would understand if you keep it as is! :ok_hand:

asolntsev commented on framework/src/play/cache/CacheImpl.java:
```diff
@@ -1,14 +1,21 @@
+  static CacheImpl instance(@Nonnull Properties playProperties) throws IOException {
+    throw new NotImplementedException();
+  }
+
```
@cies I don't think this method is needed. In Java, static methods are not inherited by subclasses.

cies replied:
@asolntsev  The `instance(Props)` method part of the singleton pattern that I want to have enforced on all implementations. I've put this method here to show to implementing devs that they need to implement this method.

Also I call this method on through the interface (and reflection) here in `Cache.java`:

```java
        // Since it implements the `CacheImpl` interface, it should have a static `instance` method.
        Method method = cacheImplClass.getDeclaredMethod("instance", Properties.class);
        cacheImpl = (CacheImpl) method.invoke(null, Play.configuration);
```

I'm not sure if it is actually needed, but it does point implementing devs in the right direction and and allow me to call it through reflection with some certainty.

Based on your comment I prefer to keep it. Maybe I do not understand it well enough though.

asolntsev replied:
> allow me to call it through reflection with some certainty

@cies For sure, this method is not needed. It doesn't add any certainty. :)

You can give a hint to developers just in javadoc.

cies replied:
Now explained with javadoc as per you request.

asolntsev commented on ehcache/test/play/cache/EhCacheImplTest.java:
```diff
@@ -63,7 +63,7 @@ public void stop() {
     } catch (IllegalStateException ignored) {
     } finally {
-      EhCacheImpl.newInstance();
+      EhCacheImpl.testInstance();
```
Why this line is needed? It doesn't have any effect.

cies replied:
@asolntsev I missed this comment. You are right, this can be removed... Will do so in a bit.

asolntsev replied:
ah yes, now I see that `EhCacheImpl` constructor is private. Then it makes sense. 
@cies But I would remove `public` modifier from `EhCacheImpl.testInstance` method. Then it will be _package private_ - this is the usual practice for methods "made available only for tests".

asolntsev commented on ehcache/src/play/cache/EhCacheImpl.java:
```diff
@@ -62,13 +64,17 @@ private EhCacheImpl() {
+  @VisibleForTesting
+  public static EhCacheImpl testInstance() {
+    return new EhCacheImpl();
   }
```
@cies Why method `testInstance()` is needed? It's simple to just call `new EhCacheImpl()` instead of `EhCacheImpl.testInstance()`.

asolntsev replied:
@cies I would make method `testInstance()` _package-private_ instead of public (it means: without "public", "protected" or "private" modifier). Then it's visible from unit-tests, and at least not visible from other packages.

cies replied:
Nice one. Fixed.

asolntsev commented on framework/src/play/cache/Cache.java:
```diff
@@ -63,17 +73,38 @@ public static void clear() {
+        Method method = cacheImplClass.getDeclaredMethod("instance", Properties.class);
+        cacheImpl = (CacheImpl) method.invoke(null, Play.configuration);
+      } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+        logger.error("Error while instantiating cache", e);
```
I would throw an exception here instead of fall-back to dummy implementation. 
The sooner user gets aware of the problem the better.

cies replied:
Changed this to throw an exception instead.

asolntsev reviewed (changes requested):
One drawback of the new approach is that now it's impossible to have both EHCache and Memcached in classpath, and select one of them by configuration property `memcached=enabled`. 

We previously used "memcached" in production and "ehcache" in tests. 
Now it will be impossible.

But for me personally, it's not a problem anymore. :)

xabolcs commented:
> One drawback of the new approach is

@cies would you mind adding a note to README about this?
Sure, the another way is to keep that `memchached=true` check.

cies commented:
> @cies would you mind adding a note to README about this? Sure, the another way is to keep that `memchached=true` check.

@xabolcs What not clear about the note I added (and you requested)?

> // or `memcached`, or disable caching by not specifying a caching

The same "note" is there for the netty3/4/javanet choice. I just followed in the vein of the existing docs

 The word `or` means either-or. Just like in the note for the netty3/4/javanet choice. If specific changes can you put here exactly what you'd like to see, or maybe propose a change by PR.

xabolcs commented on README.md:
```diff
@@ -49,23 +49,27 @@ These were not used by RePlay's users (could be reintroduced if needed).
+  implementation("io.github.replay-framework:liquibase:2.5.0")
+  implementation("io.github.replay-framework:pdf:2.5.0")
+  implementation("io.github.replay-framework:excel:2.5.0")
+  implementation("io.github.replay-framework:ehcache:2.5.0") // or `memcached`, or disable caching by not specifying a caching package
```
```suggestion
  implementation("io.github.replay-framework:ehcache:2.5.0") // or `memcached`, or disable caching by not specifying a caching package. Earlier `memcached=true` setting won't be checked anymore!
```

xabolcs commented:
@cies, thank you for pointing me to the right direction.

I tried to add a note about it.

asolntsev commented on ehcache/test/play/cache/EhCacheIntegrationTest.java:
```diff
@@ -0,0 +1,18 @@
+  public void initializationWithinRePlay() {
+    Play.configuration.setProperty("ehcache.heapSizeInEntries", "100");
+    Cache.init();
+    Cache.set("no-dummy", 1, "10s");
```
@cies  Shouldn't we reset the cache after this test?
Otherwise, this test can affect the following tests.

asolntsev commented on framework/src/play/cache/Cache.java:
```diff
@@ -63,17 +73,37 @@ public static void clear() {
-    if ("enabled".equals(Play.configuration.getProperty("memcached", "disabled"))) {
+    Class<?> memcacheImplClass = null;
+    try {
+      memcacheImplClass = Class.forName("play.cache.MemcachedImpl");
```
Nowadays, we usually implement such things using "Service Loader" mechanism (via file `META-INF/services`). 

Then this code would look simpler, something like

> CacheImpl impl = ServiceLoader.get("org.replay.Cache");
> // find any implementation of org.replay.Cache interface in all JAR files

See https://www.youtube.com/watch?v=-KGtZoFVzr8&t=11m28s

But ok, in case of Replay it's not too important. :)

cies replied:
I knew about it but never used it. I just picked --as mentioned-- the same way of implementing that was used by the dependency-pluggable webserver that RePlay ships (nett3/4/javanet).

Thanks for pointing it out.
